pinc/Project.inc

```diff
@@ -534,11 +541,12 @@ class Project
 
         $projectid = $this->projectid;
         $page_image_names = array();
-        $res = mysqli_query(DPDatabase::get_connection(), "
+        $sql = "
```

Comment:
Somewhat frustratingly, we can't rely on `$this->projectid` to be a valid project id. There are object construction paths that could mean it is anything. To ensure this is a valid project name, and not an injection attack, you can put a `validate_projectID($projectid);` before building the SQL.

Reply:
OK. It feels like I've had to scatter a bunch of validate_projectIDs around the place, which won't be great for performance. That said, they're all near DB queries, so I'm sure it's lost in the noise.

I'm not sure how `class Project` is designed, but... ordinarily wouldn't we solve this by making the validity of projectid a class invariant that's enforced by allowing only the constructor (and specific methods) to modify (and validate) $this->projectid, and making $projectid private?

Reply:
Well it's because you did it wrong. We only need them in places where it is then used as a table name in an SQL statement.

Reply:
https://www.pgdp.net/wiki/DP_Code_Best_Practices#Validating_project_tables

Reply:
Ah, I see. Removed as requested.